filesystems/blob/src/main/scala/cromwell/filesystems/blob/BlobFileSystemManager.scala

```diff
@@ -169,37 +169,64 @@ object BlobSasTokenGenerator {
     * @return A WSMBlobTokenGenerator, able to produce a valid SAS token for accessing the provided blob
     * container and endpoint that is managed by WSM
     */
-  def createBlobTokenGenerator(container: BlobContainerName,
-                               endpoint: EndpointURL,
-                               workspaceId: WorkspaceId,
+  def createBlobTokenGenerator(workspaceId: WorkspaceId,
                                containerResourceId: ContainerResourceId,
                                workspaceManagerClient: WorkspaceManagerApiClientProvider,
-                               overrideWsmAuthToken: Option[String]): BlobSasTokenGenerator = {
-    WSMBlobSasTokenGenerator(container, endpoint, workspaceId, containerResourceId, workspaceManagerClient, overrideWsmAuthToken)
+                               overrideWsmAuthToken: Option[String],
+                               buffer: Duration): BlobSasTokenGenerator = {
+    new WSMBlobSasTokenGenerator(workspaceId, containerResourceId, workspaceManagerClient, overrideWsmAuthToken, buffer)
   }
 
 }
 
-case class WSMBlobSasTokenGenerator(container: BlobContainerName,
-                                    endpoint: EndpointURL,
-                                    workspaceId: WorkspaceId,
+sealed trait SasCacheAvailable;
+case class Available(sas: AzureSasCredential) extends SasCacheAvailable;
+case class Unavailable() extends SasCacheAvailable;
+
+class WSMBlobSasTokenGenerator(workspaceId: WorkspaceId,
                                     containerResourceId: ContainerResourceId,
                                     wsmClientProvider: WorkspaceManagerApiClientProvider,
-                                    overrideWsmAuthToken: Option[String]) extends BlobSasTokenGenerator {
+                                    overrideWsmAuthToken: Option[String],
+                                    buffer: Duration) extends BlobSasTokenGenerator {
 
+  var cachedSasTokens: ConcurrentHashMap[(EndpointURL, BlobContainerName), SasCacheAvailable] = new ConcurrentHashMap[(EndpointURL, BlobContainerName), SasCacheAvailable];
```

Comment:
It seems like `WSMBlobSasTokenGenerator` is scoped to a particular container, but the cache is scoped to the BlobFilesystem. Is that right? I think it would be better to either make the `TokenGenerator` generic and always have the caller pass in container-specific bits when getting a token, or have a caching layer above/outside the `TokenGenerator`.